com.unity.render-pipelines.high-definition/Runtime/Material/Lit/Lit.hlsl

```diff
@@ -685,32 +688,31 @@ void EncodeIntoGBuffer( SurfaceData surfaceData
     }
 #endif
 
-    // RT3 - 11f:11f:10f
-    // In deferred we encode emissive color with bakeDiffuseLighting. We don't have the room to store emissiveColor.
-    // It mean that any futher process that affect bakeDiffuseLighting will also affect emissiveColor, like SSAO for example.
-#if defined(PROBE_VOLUMES_L1) || defined(PROBE_VOLUMES_L2)
+    // Random TAG which we expect will never match provided emissive or lightmap value
+    #define AO_IN_GBUFFER3_TAG float3((1 << 11), 1, (1 << 10))
 
-    if (IsUninitializedGI(builtinData.bakeDiffuseLighting))
-    {
-        // builtinData.bakeDiffuseLighting contain uninitializedGI sentinel value.
-
-        // This means probe volumes will not get applied to this pixel, only emissiveColor will.
-        // When length(emissiveColor) is much greater than length(probeVolumeOutgoingRadiance), this will visually look reasonable.
-        // Unfortunately this will break down when emissiveColor is faded out (result will pop).
-        // TODO: If evaluating probe volumes in lightloop, only write out sentinel value here, and re-render emissive surfaces.
-        // Pre-expose lighting buffer
-        outGBuffer3 = float4(all(builtinData.emissiveColor == 0.0) ? builtinData.bakeDiffuseLighting : builtinData.emissiveColor * GetCurrentExposureMultiplier(), 0.0);
-    }
-    else
-#endif
+    // RT3 - 11f:11f:10f
+    // In deferred we encode emissive color with bakeDiffuseLighting. We don't have the room to store emissiveColor separately.
+    // It mean that any futher process that affect bakeDiffuseLighting in the lightloop will also affect emissiveColor, like SSAO for example.
+    // For APV (non lightmap case) and SSGI/RTGI/Mixed bakeDiffuseLighting is 0 and below code will simply store emissiveColor
+    // Extra hack: In this last case, if emissiveColor is 0 then we store AO inside the buffer with a tag.
+    outGBuffer3 = float4(builtinData.bakeDiffuseLighting * surfaceData.ambientOcclusion + builtinData.emissiveColor, 0.0);
+    // Pre-expose lighting buffer
+    outGBuffer3.rgb *= GetCurrentExposureMultiplier();
+
+    // If this is 0 it mean that both bakeDiffuseLighting and emissiveColor are 0 and we are potentially in case of one effect
+    // so store AO instead. It doesn't matter if it is a false positive as result will be correct, this is to reduce code divergence
+    // Note: We assume that having non black baseColor * AO * lighting + Emissive is uncommon / rare (We expect mostly baseColor * AO * lighting or Emissive or baseColor * lighting + Emissive)
+    // and use this information as a tradeoff to improve quality in all others cases
+    if (all(outGBuffer3 == 0.0))
     {
-        outGBuffer3 = float4(builtinData.bakeDiffuseLighting * surfaceData.ambientOcclusion + builtinData.emissiveColor, 0.0);
-        // Pre-expose lighting buffer
-        outGBuffer3.rgb *= GetCurrentExposureMultiplier();
+        outGBuffer3.xz = AO_IN_GBUFFER3_TAG.xz;
```

Comment:
@FrancescoC-unity . I finally decided to do some kind of hack to store AO. See the comment, but short: if we don't have emissive, then we store AO. Should solve 90% of our use case and don't require extra buffer.

Reply:
Yeah sounds like a valid solution indeed